src/coreclr/vm/jitinterface.cpp

```diff
@@ -13234,207 +13253,245 @@ PCODE UnsafeJitFunctionWorker(EECodeGenManager *pJitMgr, CEECodeGenInfo *pJitInf
 
-    if (ILHeader == nullptr)
-        ILHeader = cxt.Header;
+// The reason that this is named UnsafeJitFunction is that this helper
+// method is not thread safe!  When multiple threads get in here for
+// the same pMD, ALL of them MUST return the SAME value.
+// To ensure that this happens you must call MakeJitWorker.
+// It creates a DeadlockAware list of methods being jitted and prevents us
+// from trying to jit the same method more that once.
+//
+// Calls to this method that occur to check if inlining can occur on x86,
+// are OK since they discard the return value of this method.
+PCODE UnsafeJitFunction(PrepareCodeConfig* config,
+                        _In_opt_ COR_ILMETHOD_DECODER* ILHeader,
+                        _Out_ CORJIT_FLAGS* pJitFlags,
+                        _Out_ ULONG* pSizeOfCode)
+{
+    STANDARD_VM_CONTRACT;
+    _ASSERTE(config != NULL);
+    _ASSERTE(pJitFlags != NULL);
+
+    NativeCodeVersion nativeCodeVersion = config->GetCodeVersion();
+    MethodDesc* ftn = nativeCodeVersion.GetMethodDesc();
 
     // If it's generic then we can only enter through an instantiated MethodDesc
     _ASSERTE(!ftn->IsGenericMethodDefinition());
 
-    // method attributes and signature are consistant
-    _ASSERTE(!!ftn->IsStatic() == ((methodInfo.args.callConv & CORINFO_CALLCONV_HASTHIS) == 0));
+    PCODE ret = (PCODE)NULL;
+    NormalizedTimer timer;
+    int64_t c100nsTicksInJit = 0;
+
+    COOPERATIVE_TRANSITION_BEGIN();
 
-    *pJitFlags = GetCompileFlags(config, ftn, &methodInfo);
+    LogJitMethodBegin(ftn);
+
+    timer.Start();
+
+    ULONG sizeOfILCode = 0;
 
 #ifdef FEATURE_INTERPRETER
-    bool useInterpreter = interpreterMgr->IsInterpreterLoaded();
+    InterpreterJitManager* interpreterMgr = ExecutionManager::GetInterpreterJitManager();
+    if (!interpreterMgr->IsInterpreterLoaded())
+    {
+        LPWSTR interpreterConfig;
+        IfFailThrow(CLRConfig::GetConfigValue(CLRConfig::EXTERNAL_Interpreter, &interpreterConfig));
+        if ((interpreterConfig != NULL) && !interpreterMgr->LoadInterpreter())
+        {
+            EEPOLICY_HANDLE_FATAL_ERROR_WITH_MESSAGE(COR_E_EXECUTIONENGINE, W("Failed to load interpreter"));
+        }
+    }
 
-    if (useInterpreter)
+    // If the interpreter was loaded, use it.
+    if (interpreterMgr->IsInterpreterLoaded())
     {
-        CInterpreterJitInfo interpreterJitInfo(ftn, ILHeader, interpreterMgr, !pJitFlags->IsSet(CORJIT_FLAGS::CORJIT_FLAG_NO_INLINING));
-        ret = UnsafeJitFunctionWorker(interpreterMgr, &interpreterJitInfo, pJitFlags, methodInfo, &cxt, nativeCodeVersion, pSizeOfCode);
+        CInterpreterJitInfo interpreterJitInfo{ config, ftn, ILHeader, interpreterMgr };
+        ret = UnsafeJitFunctionWorker(interpreterMgr, &interpreterJitInfo, nativeCodeVersion, pSizeOfCode, &sizeOfILCode);
+
+        // Collect the JIT flags for the interpreter
+        if (ret)
+            *pJitFlags = *interpreterJitInfo.getJitFlagsInternal();
```

Comment:
```suggestion
        ret = UnsafeJitFunctionWorker(interpreterMgr, &interpreterJitInfo, nativeCodeVersion, pSizeOfCode);

        // Collect the JIT flags for the interpreter
        if (ret)
        {
            sizeOfILCode = interpreterJitInfo.getMethodInfoInternal()->ILCodeSize;
            *pJitFlags = *interpreterJitInfo.getJitFlagsInternal();
        }
```
We can read the ILCodeSize out the same way as JIT flags.